### DISCUSSION TITLE
How to obtain "throughput" at each surface intersection in mitsuba?

### DISCUSSION THREAD

**Discussion by woAIxuexiSR** [mitsuba-renderer/mitsuba3#1401] [Q&A] [answered]
Hi there,

I’m implementing a "path guiding" algorithm in Mitsuba. I’m trying to extract the throughput at each surface intersection during path tracing using the following code:

``` python
while loop(active):
    ... # path tracing code
    thp = throughput.torch()
    valid = si.is_valid().torch()
```

However, this doesn’t seem to produce the correct results. Could you please advise on the proper way to obtain the throughput at each intersection?

Thank you in advance for your help!

**Comment by merlinND** [comment, marked as answer]
Hello @woAIxuexiSR,

You did not give much information so I am assuming that you are trying to run the latest version in JIT mode (`cuda_*` or `llvm_*` variants).
By default, Mitsuba integrators run in [_symbolic mode_](https://drjit.readthedocs.io/en/stable/cflow.html#id2), meaning that the path tracing loop gets executed in a single kernel. A great advantage is that intermediate values can be kept in registers, without constantly writing state out to global memory & reading it back in at each bounce.
But it also means that you cannot break out of the loop mid-way to start executing PyTorch code.

You can switch from symbolic mode to wavefront mode with the following flag (I think), but you should expect a big slow-down and increased memory usage:

```py
dr.set_flag(dr.JitFlag.SymbolicLoops, False)
```

See also this discussion about recording paths: https://github.com/mitsuba-renderer/mitsuba3/issues/187
